### PR TITLE
Add select values provider service documentation

### PR DESCRIPTION
**Pull request by alexander-schranz**
| Q | A
| --- | ---
| Fixed tickets | fixes #issuenum
| Related PRs | sulu/sulu#prnum
| License | MIT

#### What's in this PR?

Add select values provider service documentation.

#### Why?

Currently there is no example for a select values provider in the documentation.

### CODE REVIEW

File: cookbook/select-values-provider.rst [hunk 1]
```diff
@@ -0,0 +1,85 @@
+Create a select values provider
+===============================
+
+An example :doc:`../reference/content-types/single_select` or :doc:`../reference/content-types/select` values provider could look like the following:
```
Comment (nnatter) [inline cookbook/select-values-provider.rst:4]:
```
Sulu allows to set the value of a ``param`` of a property with the type ``expression`` via the Symfony expression language (https://symfony.com/doc/current/components/expression_language.html). 
The makes it possible to display values returned by a service in the case of the  :doc:`../reference/content-types/select` and :doc:`../reference/content-types/single_select` content-type.

.. note::

    Be aware that the provided expression is only evaluated during the initial request to the administration interface.
    If you want to provide a selection for your custom entity, you should configure the ``single_selection`` field-type
    as described in :doc:`../../book/extend-admin`.

A service for returning the values for these content types could look something like this:
```

File: cookbook/select-values-provider.rst [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    class EventTypeSelect {
+        /**
+         * @return array<int, array{name: string, title: string}>
+         */
+        public function getValues(): array
+        {
+            return [
+                [
+                    'name' => 'concert',
+                    'title' => 'Concert',
```
Comment (nnatter) [inline cookbook/select-values-provider.rst:21]: can we return a localized title?

File: cookbook/select-values-provider.rst [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+         */
+        public function getDefaultValueMultiSelect(): array
+        {
+            return [
+                ['name' => 'concert'],
+            ];
+        }
+    }
+
+The service need to be available as public service so the following need to be added to the ``config/services.yaml``:
```
Comment (nnatter) [inline cookbook/select-values-provider.rst:51]:
```
The service need to be marked as public (https://symfony.com/doc/current/service_container/alias_private.html). This is possible by adding the following lines to the ``config/services.yaml``:
```

File: cookbook/select-values-provider.rst [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+                    'name' => 'movie',
+                    'title' => 'Movie',
+                ],
+            ];
+        }
+
+        /**
+         * Optional default value for a single select.
+         */
+        public function getDefaultValue(): string
```
Comment (nnatter) [inline cookbook/select-values-provider.rst:33]:
```suggestion
        public function getSingleSelectDefaultValue(): string
```

File: cookbook/select-values-provider.rst [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        {
+            return 'concert';
+        }
+
+        /**
+         * Optional default value for a multi select.
+         *
+         * @return array<int, array{name: string}>
+         */
+        public function getDefaultValueMultiSelect(): array
```
Comment (nnatter) [inline cookbook/select-values-provider.rst:43]:
```suggestion
        public function getMultiSelectDefaultValue(): array
```

File: cookbook/select-values-provider.rst [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+The service need to be available as public service so the following need to be added to the ``config/services.yaml``:
+
+.. code-block:: yaml
+
+    App\Content\Select\:
+        resource: '../src/Content/Select/'
+        public: true
+
+Then you can use the value provider the following way in your pages template:
```
Comment (nnatter) [inline cookbook/select-values-provider.rst:59]:
```suggestion
When your service is marked as public, you can access it in your template like this:
```

File: reference/content-types/select.rst [hunk 7]
```diff
@@ -76,11 +76,13 @@ You can use symfony expression language to access values or set a default value
         </meta>
```
Comment (nnatter) [inline reference/content-types/select.rst:76]: Would propose the following:
```
Service Example
---------------

By using a ``param`` with the type ``expression``, it is also possible to display the values returned by a service (:doc:`../../cookbook/select-values-provider):

.. note::

    Be aware that the provided expression is only evaluated during the initial request to the administration interface.
    If you want to provide a selection for your custom entity, you should configure the ``single_selection`` field-type
    as described in :doc:`../../book/extend-admin`.

.. code-block:: xml

    <property name="single" type="single_select">
        <meta>
            <title lang="en">Single Select</title>
        </meta>

        <params>
            <param name="default_value" type="expression" value="service('App\\Content\\Select\\EventTypeSelect').getDefaultValue()"/>
            <param name="values" type="expression" value="service('App\\Content\\Select\\EventTypeSelect').getValues()"/>
        </params>
    </property>
```

File: reference/content-types/single_select.rst [hunk 8]
```diff
@@ -67,11 +67,13 @@ You can use symfony expression language to access values or set a default value
...
 
         <params>
-            <param name="default_value" type="expression" value="service('your_service').getDefaultValue()"/>
-            <param name="values" type="expression" value="service('your_service').getValues()"/>
+            <param name="default_value" type="expression" value="service('App\\Content\\Select\\EventTypeSelect').getDefaultValue()"/>
+            <param name="values" type="expression" value="service('App\\Content\\Select\\EventTypeSelect').getValues()"/>
         </params>
     </property>
 
+To create the service have a look at :doc:`../../cookbook/select-values-provider`.
```
Comment (nnatter) [inline reference/content-types/single_select.rst:75]: same as above 🙂

File: cookbook/select-values-provider.rst (hunk 2 above)
Comment (alexander-schranz) [reply cookbook/select-values-provider.rst:21]: Not sure I never tried it and it seems the expression language does currently not support getting the `locale`.

File: reference/content-types/single_select.rst [hunk 9]
```diff
@@ -61,14 +65,14 @@ You can use symfony expression language to access values or set a default value
 
 .. code-block:: xml
 
-    <property name="single" type="single_select">
+    <property name="list" type="single_select">
```
Comment (nnatter) [inline reference/content-types/single_select.rst:68]:
```suggestion
    <property name="single" type="single_select">
```

File: reference/content-types/single_select.rst [hunk 10]
```diff
@@ -51,7 +51,11 @@ Example
         </params>
     </property>
 
-You can use symfony expression language to access values or set a default value from a service.
+Expression Example
+------------------
+
+By u sing a ``param`` with the type ``expression``, it is also possible to display
```
Comment (nnatter) [inline reference/content-types/single_select.rst:57]:
```suggestion
By using a ``param`` with the type ``expression``, it is also possible to display
```

File: cookbook/select-values-service.rst [hunk 11]
```diff
@@ -0,0 +1,97 @@
+Display values from a service in a select
+=========================================
+
+Sulu allows to set the value of a ``param`` of a property with the type ``expression``
+via the Symfony expression language (https://symfony.com/doc/current/components/expression_language.html).
```
Comment (nnatter) [inline cookbook/select-values-service.rst]: can we make `Symfony expression language ` a link

File: cookbook/select-values-service.rst [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+         */
+        public function getMultiSelectDefaultValue(): array
+        {
+            return [
+                ['name' => 'concert'],
+            ];
+        }
+    }
+
+The service need to be marked as public (https://symfony.com/doc/current/service_container/alias_private.html).
```
Comment (nnatter) [inline cookbook/select-values-service.rst]: can we make `marked as public ` a link

File: cookbook/select-values-service.rst [hunk 13]
```diff
@@ -0,0 +1,97 @@
+Display values from a service in a select
+=========================================
+
+Sulu allows to set the value of a ``param`` of a property with the type ``expression``
+via the Symfony expression language (https://symfony.com/doc/current/components/expression_language.html).
+The makes it possible to display values returned by a service in the case of the
```
Comment (nnatter) [inline cookbook/select-values-service.rst:6]:
```suggestion
This makes it possible to display values returned by a service in the case of the
```

File: reference/content-types/select.rst [hunk 14]
```diff
@@ -60,13 +60,17 @@ Example
         </params>
     </property>
 
-You can use symfony expression language to access values or set a default value from a service.
+Expression Example
+------------------
+
+By u sing a ``param`` with the type ``expression``, it is also possible to display
```
Comment (nnatter) [inline reference/content-types/select.rst:66]:
```suggestion
By using a ``param`` with the type ``expression``, it is also possible to display
```

File: cookbook/select-values-service.rst [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+
+Sulu allows to set the value of a ``param`` of a property with the type ``expression``
+via the `Symfony expression language`_.
+The makes it possible to display values returned by a service in the case of the
+:doc:`../reference/content-types/select` and :doc:`../reference/content-types/single_select` content-type.
+
+.. note::
+
+    Be aware that the provided expression is only evaluated during the initial request to the administration interface.
+    If you want to provide a selection for your custom entity, you should configure the ``single_selection`` field-type
```
Comment (nnatter) [inline cookbook/select-values-service.rst:12]:
```suggestion
    If you want to provide a selection for your custom entity, you should configure the ``selection`` or ``single_selection`` field-type
```

File: reference/content-types/select.rst [hunk 16]
```diff
@@ -60,13 +60,17 @@ Example
...
+the values returned by a service (:doc:`../../cookbook/select-values-service`).
 
 .. note::
 
     Be aware that the provided expression is only evaluated during the initial request to the administration interface.
-    If you want to provide a selection for your custom entity, you should configure the ``selection`` field-type as
-    described in :doc:`../../book/extend-admin`.
+    If you want to provide a selection for your custom entity, you should configure the ``single_selection`` field-type
```
Comment (nnatter) [inline reference/content-types/select.rst:72]:
```suggestion
    If you want to provide a selection for your custom entity, you should configure the ``selection`` field-type
```

**Review by nnatter — APPROVED** [review]

**Pull request merged**